common/src/cost_model.rs

```diff
@@ -0,0 +1,465 @@
+            WHERE deployment != 'global'
+            ORDER BY deployment ASC
+            "#
+        )
+        .fetch_all(pool)
+        .await?
+    } else {
+        sqlx::query_as!(
+            DbCostModel,
+            r#"
+            SELECT deployment, model, variables
+            FROM "CostModels"
+            WHERE deployment = ANY($1)
+            AND deployment != 'global'
+            ORDER BY deployment ASC
+            "#,
+            &hex_ids
+        )
+        .fetch_all(pool)
+        .await?
+    }
+    .into_iter()
+    .map(CostModel::try_from)
+    .collect::<Result<Vec<_>, _>>()?;
+
+    let deployments_with_models = models
+        .iter()
+        .map(|model| &model.deployment)
+        .collect::<HashSet<_>>();
+
+    let deployments_without_models = deployments
+        .iter()
+        .filter(|deployment| !deployments_with_models.contains(deployment))
+        .collect::<HashSet<_>>();
+
+    // Query the global cost model
+    if let Some(global_model) = global_cost_model(pool).await? {
+        // For all deployments that have a cost model, merge it with the global one
+        models = models
+            .into_iter()
+            .map(|model| merge_global(model, &global_model))
+            // Inject a cost model for all deployments that don't have one
+            .chain(
+                deployments_without_models
+                    .into_iter()
+                    .map(|deployment| CostModel {
+                        deployment: deployment.to_owned(),
+                        model: global_model.model.clone(),
+                        variables: global_model.variables.clone(),
+                    }),
+            )
+            .collect();
+    }
+
+    Ok(models)
+}
+
+/// Make database query for a cost model indexed by deployment id
+pub async fn cost_model(
+    pool: &PgPool,
+    deployment: &DeploymentId,
+) -> Result<Option<CostModel>, anyhow::Error> {
+    let model = sqlx::query_as!(
+        DbCostModel,
+        r#"
+        SELECT deployment, model, variables
+        FROM "CostModels"
+        WHERE deployment = $1
+        AND deployment != 'global'
+        "#,
+        format!("{:#x}", deployment),
+    )
+    .fetch_optional(pool)
+    .await?
+    .map(CostModel::try_from)
+    .transpose()?;
+
+    let global_model = global_cost_model(pool).await?;
+
+    Ok(match (model, global_model) {
+        // If we have no global model, return whatever we can find for the deployment
+        (None, None) => None,
+        (Some(model), None) => Some(model),
+
+        // If we have a cost model and a global cost model, merge them
+        (Some(model), Some(global_model)) => Some(merge_global(model, &global_model)),
+
+        // If we have only a global model, use that
+        (None, Some(global_model)) => Some(CostModel {
+            deployment: deployment.to_owned(),
+            model: global_model.model,
+            variables: global_model.variables,
+        }),
```

Comment:
I'm curious what made you choose the order here. Is `None, None` the most likely path?

Reply:
TBH, I'm just moving this code to common, it was previously on `service/src/database.rs`